cpu/stm32f4/periph/i2c.c

```diff
@@ -494,32 +419,31 @@ void i2c_poweroff(i2c_t dev)
 
 static void _start(I2C_TypeDef *dev, uint8_t address, uint8_t rw_flag)
 {
-    /* wait for device to be ready */
-    DEBUG("Wait for device to be ready\n");
-
-    while (dev->SR2 & I2C_SR2_BUSY) {}
+start:
 
     /* generate start condition */
-    DEBUG("Generate start condition\n");
     dev->CR1 |= I2C_CR1_START;
-    DEBUG("Wait for SB flag to be set\n");
 
+    /* Wait for SB flag to be set */
     while (!(dev->SR1 & I2C_SR1_SB)) {}
 
     /* send address and read/write flag */
-    DEBUG("Send address\n");
     dev->DR = (address << 1) | rw_flag;
-    /* clear ADDR flag by reading first SR1 and then SR2 */
-    DEBUG("Wait for ADDR flag to be set\n");
 
-    while (!(dev->SR1 & I2C_SR1_ADDR)) {}
+    /* Wait for ADDR flag to be set */
+    while (!(dev->SR1 & I2C_SR1_ADDR)) {
+        if (dev->SR1 & I2C_SR1_AF) {
+            /* if the device answers NACK on sending the address, retry */
```

Comment:
I think the better solution would be to return an error code on address NACK. This prevents lockup if trying to talk to non-existent devices...